File: leetcode/637-average-of-levels-in-binary-tree/main.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
	bfs, iterative

	Time	O(n)
	Space	O(h)
	12ms beats 100%
*/
func averageOfLevels(root *TreeNode) []float64 {
	if root == nil {
		return []float64{}
	}
	var result []float64
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		lengthOnSameLevel := len(queue)
		sum := 0.0
		i := 0
		for ; i < lengthOnSameLevel; i++ {
			head := queue[0]
			queue = queue[1:len(queue)]
			sum += float64(head.Val)
			if head.Left != nil {
				queue = append(queue, head.Left)
			}
			if head.Right != nil {
				queue = append(queue, head.Right)
			}
		}
		avg := sum / float64(i)
		result = append(result, avg)
	}
	return result
}

func main() {

}

File: leetcode/637-average-of-levels-in-binary-tree/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

"""
    1st: bfs

    Time    O(N)
    Space   O(N)
    52 ms, faster than 74.26%
"""


class Solution:
    def averageOfLevels(self, root: TreeNode) -> List[float]:
        if root == None:
            return []
        res = []
        q = [root]
        while len(q) > 0:
            n = len(q)
            total = 0
            for _ in range(n):
                node = q.pop(0)
                total += node.val
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            res.append(total/n)
        return res
